File: leetsync/1917-maximum-average-pass-ratio/maximum-average-pass-ratio.go
type element struct {
	avg   float64
	index int
}

type pq []*element

func (c pq) Len() int {
	return len(c)
}

func (c pq) Less(i, j int) bool {
	return c[i].avg > c[j].avg
}

func (c pq) Swap(i, j int) {
	c[i], c[j] = c[j], c[i]
}

func (c *pq) Push(x interface{}) {
	*c = append(*c, x.(*element))
}

func (c *pq) Pop() interface{} {
	n := len(*c)
	x := (*c)[n-1]
	*c = (*c)[:n-1]
	return x
}

func calAvg(c []int) float64 {
	return float64(c[0]+1)/float64(c[1]+1) - float64(c[0])/float64(c[1])
}

func maxAverageRatio(classes [][]int, extraStudents int) float64 {
	h := pq{}
	for i, c := range classes {
		heap.Push(&h, &element{
			avg:   calAvg(c),
			index: i,
		})
	}
	for i := 0; i < extraStudents; i++ {
		c := heap.Pop(&h).(*element)
		classes[c.index][0]++
		classes[c.index][1]++
		c.avg = calAvg(classes[c.index])
		heap.Push(&h, c)
	}
	var sum float64
	for _, c := range classes {
		sum += float64(c[0]) / float64(c[1])
	}
	return sum / float64(len(classes))
}

File: leetsync/1917-maximum-average-pass-ratio/maximum-average-pass-ratio.java
class Solution {
    public double maxAverageRatio(int[][] classes, int extraStudents) {
        int n = classes.length;
        // Priority queue to act as a max-heap, storing pairs of {max-delta, index}
        PriorityQueue<double[]> pq = new PriorityQueue<>((a, b) -> Double.compare(b[0], a[0]));

        // Initialize the priority queue with the delta values and indices
        for (int i = 0; i < n; i++) {
            double currentPR = (double) classes[i][0] / classes[i][1];
            double newPR = (double) (classes[i][0] + 1) / (classes[i][1] + 1);
            double delta = newPR - currentPR;
            pq.add(new double[] { delta, i });
        }

        // Allocate extra students
        while (extraStudents-- > 0) {
            // Extract the class with the maximum delta
            double[] current = pq.poll();
            int index = (int) current[1];
            // Update the class with an extra student
            classes[index][0]++;
            classes[index][1]++;
            // Recalculate the delta for this class
            double currentPR = (double) classes[index][0] / classes[index][1];
            double newPR = (double) (classes[index][0] + 1) / (classes[index][1] + 1);
            double delta = newPR - currentPR;
            // Push the updated delta and index back into the priority queue
            pq.add(new double[] { delta, index });
        }
        // Calculate the final average pass ratio
        double result = 0.0;
        for (int[] clazz : classes) {
            result += (double) clazz[0] / clazz[1];
        }
        return result / n;
    }
}